@QueryResult leads to late class cast exception in some cases

Spring Data Neo4j **5.1**, Neo4j OGM **3.1.4**

Given a query result like this

```
@QueryResult
public class ThingResult {
	private String something;

	public String getSomething() {
		return something;
	}

	public void setSomething(String something) {
		this.something = something;
	}

	private List<AnotherThing> things;

	public List<AnotherThing> getThings() {
		return things;
	}

	public void setThings(List<AnotherThing> things) {
		this.things = things;
	}
}
```

That uses another class

```
public class AnotherThing  {

	private String name;
	
	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}
}
```

and a query like 

```
public interface ThingRepository extends Neo4jRepository<ThingEntity, Long> {
	@Query("MATCH (t:ThingEntity) RETURN 'das ist ein Test' as something, collect({name: t.name}) as things")
	List<ThingResult> findResults();
}
```

I'll end up with a class cast exception:

```
Caused by: java.lang.ClassCastException: class java.util.Collections$SingletonMap cannot be cast to class com.example.demo.AnotherThing (java.util.Collections$SingletonMap is in module java.base of loader 'bootstrap'; com.example.demo.AnotherThing is in unnamed module of loader 'app')
	at com.example.demo.DemoApplication.run(DemoApplication.java:33) ~[classes/:na]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:813) ~[spring-boot-2.1.0.RELEASE.jar:2.1.0.RELEASE]
	... 10 common frames omitted
```

This can be reproduced in some variants. List of entities in query result works most of the time, however if the result in the query doesn't contain ids, the native id cache is triggered anyway and throws a NPE.

This is maybe only a documentation issue and I'm not suggesting we should support this (even though I prefer having it), but throwing this exception as late as after query execution not good.

Attached is a reproducer.
[queryresult.zip](https://github.com/neo4j/neo4j-ogm/files/2571856/queryresult.zip)

Related #391, possible #252, i had similar errors in between.

--- a/core/src/main/java/org/neo4j/ogm/context/GraphEntityMapper.java
+++ b/core/src/main/java/org/neo4j/ogm/context/GraphEntityMapper.java
@@ -294,15 +294,12 @@ public class GraphEntityMapper {
             Object value = property.getValue();
             // merge iterable / arrays and co-erce to the correct attribute type
             if (writer.type().isArray() || Iterable.class.isAssignableFrom(writer.type())) {
-                FieldInfo reader = classInfo.getFieldInfo(property.getKey().toString());
-                if (reader != null) {
-                    Class<?> paramType = writer.type();
-                    Class elementType = underlyingElementType(classInfo, property.getKey().toString());
-                    if (paramType.isArray()) {
-                        value = EntityAccessManager.merge(paramType, value, new Object[] {}, elementType);
-                    } else {
-                        value = EntityAccessManager.merge(paramType, value, Collections.emptyList(), elementType);
-                    }
+                Class<?> paramType = writer.type();
+                Class elementType = underlyingElementType(classInfo, property.getKey().toString());
+                if (paramType.isArray()) {
+                    value = EntityAccessManager.merge(paramType, value, new Object[] {}, elementType);
+                } else {
+                    value = EntityAccessManager.merge(paramType, value, Collections.emptyList(), elementType);
                 }
             }
             writer.write(instance, value);

--- a/core/src/main/java/org/neo4j/ogm/context/SingleUseEntityMapper.java
+++ b/core/src/main/java/org/neo4j/ogm/context/SingleUseEntityMapper.java
@@ -18,21 +18,24 @@
  */
 package org.neo4j.ogm.context;
 
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 
 import org.neo4j.ogm.exception.core.MappingException;
 import org.neo4j.ogm.metadata.ClassInfo;
+import org.neo4j.ogm.metadata.DescriptorMappings;
 import org.neo4j.ogm.metadata.FieldInfo;
 import org.neo4j.ogm.metadata.MetaData;
-import org.neo4j.ogm.metadata.DescriptorMappings;
 import org.neo4j.ogm.metadata.reflect.EntityAccessManager;
 import org.neo4j.ogm.metadata.reflect.EntityFactory;
 import org.neo4j.ogm.model.RowModel;
 import org.neo4j.ogm.session.EntityInstantiator;
+import org.neo4j.ogm.support.ClassUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -42,6 +45,7 @@ import org.slf4j.LoggerFactory;
  *
  * @author Adam George
  * @author Luanne Misquitta
+ * @author Michael J. Simons
  */
 public class SingleUseEntityMapper {
 
@@ -64,8 +68,8 @@ public class SingleUseEntityMapper {
     /**
      * Constructs a new {@link SingleUseEntityMapper} based on the given mapping {@link MetaData}.
      *
-     * @param mappingMetaData The {@link MetaData} to use for performing mappings
-     * @param entityInstantiator   The entity factory to use.
+     * @param mappingMetaData    The {@link MetaData} to use for performing mappings
+     * @param entityInstantiator The entity factory to use.
      */
     public SingleUseEntityMapper(MetaData mappingMetaData, EntityInstantiator entityInstantiator) {
         this.metadata = mappingMetaData;
@@ -114,7 +118,6 @@ public class SingleUseEntityMapper {
             ".  At present, only @Result types that are discovered by the domain entity package scanning can be mapped.");
     }
 
-    // TODO: the following is all pretty much identical to GraphEntityMapper so should probably be refactored
     private void writeProperty(ClassInfo classInfo, Object instance, Map.Entry<String, Object> property) {
 
         FieldInfo writer = classInfo.getFieldInfo(property.getKey());
@@ -126,36 +129,83 @@ public class SingleUseEntityMapper {
             }
         }
 
-        if (writer == null && property.getKey().equals(
-            "id")) { //When mapping query results to objects that are not domain entities, there's no concept of a GraphID
-            FieldInfo idField = classInfo.identityField();
-            if (idField != null) {
-                writer = idField;
-            }
-        }
+        if (writer == null) {
+            logger.debug("Unable to find property: {} on class: {} for writing", property.getKey(), classInfo.name());
+        } else {
+            Class elementType = DescriptorMappings.getType(writer.getTypeDescriptor());
+            boolean targetIsCollection = writer.type().isArray() || Iterable.class.isAssignableFrom(writer.type());
 
-        if (writer != null) {
-            Object value = property.getValue();
-            if (value != null && value.getClass().isArray()) {
-                value = Arrays.asList((Object[]) value);
-            }
-            if (writer.type().isArray() || Iterable.class.isAssignableFrom(writer.type())) {
-                Class elementType = underlyingElementType(classInfo, property.getKey());
-                value = writer.type().isArray()
-                    ? EntityAccessManager.merge(writer.type(), value, new Object[] {}, elementType)
-                    : EntityAccessManager.merge(writer.type(), value, Collections.EMPTY_LIST, elementType);
+            Object value = mapKnownNestedClasses(elementType, property.getKey(), property.getValue(),
+                targetIsCollection);
+
+            // merge iterable / arrays and co-erce to the correct attribute type
+            if (targetIsCollection) {
+                if (value == null) {
+                    value = Collections.emptyList();
+                } else if (value.getClass().isArray()) {
+                    value = Arrays.asList((Object[]) value);
+                }
+
+                Class<?> paramType = writer.type();
+                if (paramType.isArray()) {
+                    value = EntityAccessManager.merge(paramType, value, new Object[] {}, elementType);
+                } else {
+                    value = EntityAccessManager.merge(paramType, value, Collections.emptyList(), elementType);
+                }
             }
             writer.write(instance, value);
-        } else {
-            logger.warn("Unable to find property: {} on class: {} for writing", property.getKey(), classInfo.name());
         }
     }
 
-    private Class underlyingElementType(ClassInfo classInfo, String propertyName) {
-        FieldInfo fieldInfo = classInfo.propertyField(propertyName);
-        if (fieldInfo != null) {
-            return DescriptorMappings.getType(fieldInfo.getTypeDescriptor());
+    /**
+     * @param elementType  The target type, must not be null
+     * @param property     The name of the property
+     * @param value        The value (can be null)
+     * @param asCollection whether to create a collection or not
+     * @return The mapped value
+     */
+    Object mapKnownNestedClasses(Class elementType, String property, Object value, boolean asCollection) {
+
+        Object mappedValue = value;
+        if (metadata.classInfo(elementType) != null) {
+            List<Object> nestedObjects = new ArrayList<>();
+
+            for (Object nestedPropertyMap : iterableOf(value)) {
+                if (nestedPropertyMap instanceof Map) {
+                    // Recursively map maps
+                    nestedObjects.add(map(elementType, (Map<String, Object>) nestedPropertyMap));
+                } else if (elementType.isInstance(nestedPropertyMap) || ClassUtils.isEnum(elementType)) {
+                    // Add fitting types and enums directly
+                    nestedObjects.add(nestedPropertyMap);
+                } else {
+                    logger.warn("Cannot map {} to a nested result object for property {}", nestedPropertyMap,
+                        property);
+                }
+            }
+            if (asCollection) {
+                mappedValue = nestedObjects;
+            } else if (nestedObjects.isEmpty()) {
+                mappedValue = Collections.emptyList();
+            } else if (nestedObjects.size() == 1) {
+                mappedValue = nestedObjects.get(0);
+            } else {
+                logger.warn(
+                    "Cannot map property {} from result set: The result contains more than one entry for the property.",
+                    property);
+            }
         }
-        return classInfo.getUnderlyingClass();
+        return mappedValue;
+    }
+
+    Iterable iterableOf(Object thingToIterator) {
+        if (thingToIterator == null) {
+            return Collections.emptyList();
+        } else if (thingToIterator instanceof Iterable) {
+            return ((Iterable) thingToIterator);
+        } else if (thingToIterator.getClass().isArray()) {
+            return Arrays.asList((Object[]) thingToIterator);
+        } else {
+            return Collections.singletonList(thingToIterator);
+        }
     }
 }

--- a/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/context/SingleUseEntityMapperTest.java
+++ b/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/context/SingleUseEntityMapperTest.java
@@ -18,51 +18,86 @@
  */
 package org.neo4j.ogm.context;
 
+import static java.util.Collections.*;
+import static org.assertj.core.api.Assertions.*;
+
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
 
+import org.junit.BeforeClass;
 import org.junit.Test;
+import org.neo4j.ogm.domain.gh551.AnotherThing;
+import org.neo4j.ogm.domain.gh551.ThingResult;
 import org.neo4j.ogm.metadata.MetaData;
 import org.neo4j.ogm.metadata.reflect.ReflectionEntityInstantiator;
+import org.neo4j.ogm.session.SessionFactory;
+import org.neo4j.ogm.testutil.TestContainersTestBase;
 
 /**
- * TODO: This test doesn't assert anything.
- *
- * @author vince
+ * @author Vince Bickers
+ * @author Michael J. Simons
  */
-public class SingleUseEntityMapperTest {
+public class SingleUseEntityMapperTest extends TestContainersTestBase {
 
-    private MetaData metaData = new MetaData("org.neo4j.ogm.context");
+    private static SessionFactory sessionFactory;
 
-    //	@Test
-    //	public void shouldMapFromRowModel() throws Exception {
-    //
-    //
-    //	}
+    @BeforeClass
+    public static void oneTimeSetUp() {
+        sessionFactory = new SessionFactory(getDriver(), "org.neo4j.ogm.domain.gh551");
+
+        // Prepare test data
+        sessionFactory.openSession()
+            .query("unwind range(1,10) as x with x create (n:ThingEntity {name: 'Thing ' + x}) return n", EMPTY_MAP);
+    }
+
+    @Test
+    public void singleUseEntityMapperShouldWorkWithNestedObjects() {
+
+        SingleUseEntityMapper entityMapper =
+            new SingleUseEntityMapper(sessionFactory.metaData(),
+                new ReflectionEntityInstantiator(sessionFactory.metaData()));
+
+        Iterable<Map<String, Object>> results = sessionFactory.openSession()
+            .query("MATCH (t:ThingEntity) RETURN 'a name' as something, collect({name: t.name}) as things", EMPTY_MAP)
+            .queryResults();
+
+        assertThat(results).hasSize(1);
+
+        ThingResult thingResult = entityMapper.map(ThingResult.class, results.iterator().next());
+        assertThat(thingResult.getSomething()).isEqualTo("a name");
+        assertThat(thingResult.getThings())
+            .hasSize(10)
+            .extracting(AnotherThing::getName)
+            .allSatisfy(s -> s.startsWith("Thing"));
+    }
 
     @Test
     public void shouldMapFromMap() {
 
+        MetaData metaData = new MetaData("org.neo4j.ogm.context");
+        SingleUseEntityMapper entityMapper = new SingleUseEntityMapper(metaData,
+            new ReflectionEntityInstantiator(metaData));
+
         Collection<Object> toReturn = new ArrayList<>();
-        SingleUseEntityMapper entityMapper = new SingleUseEntityMapper(metaData, new ReflectionEntityInstantiator(metaData));
 
         Iterable<Map<String, Object>> results = getQueryResults();
-
         for (Map<String, Object> result : results) {
             toReturn.add(entityMapper.map(UserResult.class, result));
         }
+
+        assertThat(toReturn).hasSize(1);
+        assertThat(toReturn).first().isInstanceOf(UserResult.class);
+        UserResult userResult = (UserResult) toReturn.iterator().next();
+        assertThat(userResult.getProfile()).containsAllEntriesOf(
+            (Map<? extends String, ?>) results.iterator().next().get("profile"));
     }
 
     private Iterable<Map<String, Object>> getQueryResults() {
-        List<Map<String, Object>> results = new ArrayList<>();
-
-        Map<String, Object> result = new HashMap<>();
 
         Map<String, Object> profile = new HashMap<>();
-
         profile.put("enterpriseId", "p.enterpriseId");
         profile.put("clidUuid", "u.clidUuid");
         profile.put("profileId", "p.clidUuid");
@@ -72,9 +107,6 @@ public class SingleUseEntityMapperTest {
         profile.put("roles", "roles");
         profile.put("connectionType", "connectionType");
 
-        result.put("profile", profile);
-        results.add(result);
-
-        return results;
+        return Collections.singletonList(Collections.singletonMap("profile", profile));
     }
 }

--- a/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/domain/gh551/AnotherThing.java
+++ b/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/domain/gh551/AnotherThing.java
@@ -1,0 +1,35 @@
+/*
+ * Copyright (c) 2002-2019 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.ogm.domain.gh551;
+
+/**
+ * @author Michael J. Simons
+ */
+public class AnotherThing  {
+
+	private String name;
+
+	public String getName() {
+		return name;
+	}
+
+	public void setName(String name) {
+		this.name = name;
+	}
+}

--- a/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/domain/gh551/ThingEntity.java
+++ b/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/domain/gh551/ThingEntity.java
@@ -1,0 +1,58 @@
+/*
+ * Copyright (c) 2002-2019 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.ogm.domain.gh551;
+
+import org.neo4j.ogm.annotation.GeneratedValue;
+import org.neo4j.ogm.annotation.Id;
+import org.neo4j.ogm.annotation.NodeEntity;
+
+/**
+ * @author Michael J. Simons
+ */
+@NodeEntity
+public class ThingEntity {
+	@Id @GeneratedValue
+	private Long id;
+
+	private String name;
+
+    public ThingEntity() {
+    }
+
+    public ThingEntity(String name) {
+		this.name = name;
+	}
+
+	public long getId() {
+		return id;
+	}
+
+	public void setId(long id) {
+		this.id = id;
+	}
+
+	public String getName() {
+		return name;
+	}
+
+	public void setName(String name) {
+		this.name = name;
+	}
+}
+

--- a/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/domain/gh551/ThingResult.java
+++ b/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/domain/gh551/ThingResult.java
@@ -1,0 +1,46 @@
+/*
+ * Copyright (c) 2002-2019 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.ogm.domain.gh551;
+
+import java.util.List;
+
+/**
+ * @author Michael J. Simons
+ */
+public class ThingResult {
+	private String something;
+
+	public String getSomething() {
+		return something;
+	}
+
+	public void setSomething(String something) {
+		this.something = something;
+	}
+
+	private List<AnotherThing> things;
+
+	public List<AnotherThing> getThings() {
+		return things;
+	}
+
+	public void setThings(List<AnotherThing> things) {
+		this.things = things;
+	}
+}

